README.md
# Aloodak
Aloodak is a Telegram bot that reports the air pollution status in various cities in the world. The version specific to Tehran, Iran is available at 
[@Aloodak](https://t.me/Aloodak). 

## Setup
To run Aloodak, you need a Telegram Bot ([Create one here](https://t.me/BotFather)) a channel and an API key from IQAir ([Get one here](https://www.iqair.com/dashboard/api)). Then, you need to create a `config.conf.json` file in Aloodak's root directory.

```bash
git clone https://github.com/frowzyispenguin/aloodak.git
cd aloodak
python3 -m venv .env
source .env/bin/activate
pip3 install -r ./requirements.txt
cp config.example.json config.conf.json
editor config.conf.json
```

## Usage
You can use a cronjob to run the bot every n minutes. The main instance of Aloodak runs once every 30 minutes. Please note that Aloodak won't post a new post to your channel if the measurement doesn't change from last time it ran. It will also send all its messages in silenced mode. (No notifications will be sent)

```bash
30 * * * * /PATH/TO/aloodak/.env/bin/python3 /PATH/TO/aloodak/main.py
```

## Contributions
Aloodak is always evolving and pull requests are always appreciated. if you encounter a bug, don't hesitate to make an issue or better yet, a pull request!

## License & DMCA
Unless otherwise noted, this project is licensed under the GNU GPLv3 license. See the accompanying LICENSE file for more information.

We take any DMCA requests very seriously, so if any of our assets are violating your intellectual properties, don't hesitate to contact us via frowzyispenguin[at]riseup.net

"AirVisual", "IQAir", and "Telegram" are trademarks of their respective owners.

CHANGELOG.md
## 2020-01-11 Aloodak Renewed
Aloodak is now refactored!

* Aloodak now uses the official IQAir API instead of Web Scraping
* Configuring Aloodak is now done using the `config.conf.json` file
* Aloodak now uses CRON to run instead of an infinite loop
* Aloodak now stores its generated content in memory instead of saving them as files.
* PM 2.5 was dropped out of Aloodak due to API limitations
* Dropped use of Pyrogram and BS4 in favor of requests


core.py
""" Aloodak
    Copyright (C) 2021  Frowzy et al

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.

    For copyright related issues, contact frowzyispenguin<at>riseup.net
"""
from io import BytesIO
from PIL import Image, ImageDraw, ImageFont
import requests
from misc import now, today, toPersianNumerics

statusbar = {
    "Hazardous": "خطر اضطراری",
    "Very Unhealthy": "خیلی ناسالم",
    "Unhealthy": "ناسالم",
    "Unhealthy for Sensitive Groups": "ناسالم برای گروه های حساس",
    "Moderate": "سالم",
    "Good": "پاک"
}

# defining textcolor / light/dark colors has calculated by hps
color = {
    "Hazardous": (104,62,81),
    "Very Unhealthy": (99,70,117),
    "Unhealthy": (175,44,59),
    "Unhealthy for Sensitive Groups": (178,88,38),
    "Moderate": (165,127,35),
    "Good": (113,139,58)
}


def aqiToStatus(aqi: int):
    scale = [
        {'title': 'Hazardous', 'min': 300},
        {'title': 'Very Unhealthy', 'min': 201},
        {'title': 'Unhealthy', 'min': 151},
        {'title': 'Unhealthy for Sensitive Groups', 'min': 101},
        {'title': 'Moderate', 'min': 51},
        {'title': 'Good',  'min': 0},
    ]

    for status in scale:
        if aqi >= status['min']:
            return status['title']


def getPollutionData(iqair_key, country, province, city):
    url = f"http://api.airvisual.com/v2/city?city={city}&state={province}&country={country}&key={iqair_key}"
    json_results = requests.get(url).json() 
    # TODO write a fail-safe mechanism to stop the bot and log the incident.
    return json_results['data']


class Visualizor():
    def __init__(self, data: dict, manualCall=False):
        
        self.font = ImageFont.truetype("assets/fonts/Sahel-Black.ttf", 140) # rate font style

        
        self.temperature = data['current']['weather']['tp']
        self.humidity = data['current']['weather']['hu']
        self.aqi = data['current']['pollution']['aqius']
        self.status = aqiToStatus(self.aqi)
        self.status_text = statusbar[aqiToStatus(self.aqi)]
        self.color = color[self.status]

        self.imgdir = f"assets/badges/{self.status}.png"

        if not manualCall:
            self.generateCaption()
            self.generateImage()

    def generateImage(self): 
        self.image = Image.open(self.imgdir) # opening image
        self.draw_object = ImageDraw.Draw(self.image) # loafing draw module 
        
        # Adjust the position based on the number of digits
        if self.aqi >= 100:
            position = 100
        elif self.aqi >= 10:
            position = 130
        else:
            position = 150

        self.draw_object.text((position,100), toPersianNumerics(self.aqi), self.color, font=self.font) 
        
        # self.image.save("report.png")


        image_file = BytesIO()
        self.image.save(image_file, "PNG")
        image_file.seek(0)

        self.imageb = image_file
        return image_file

    def generateCaption(self):
        items = {"🔹شاخص آلودگی هوا : " : str(self.aqi),
                 "🔹وضعیت هوا : " : self.status_text,
                 "💧 رطوبت هوا : ": str(self.humidity),
                 "🌡 دمای هوا : ": str(self.temperature),
                 "🗓تاریخ : " : ' - '.join([x for x in today()]),
                 "🕓ساعت : " : now()
                }

        caption = [(item+items[item]) for item in items]
        caption = '\n'.join(x for x in caption)
        self.caption = toPersianNumerics(caption)
        
        return self.caption


main.py
""" Aloodak
    Copyright (C) 2021  Frowzy et al

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.

    For copyright related issues, contact frowzyispenguin<at>riseup.net
"""
import argparse
from core import getPollutionData, Visualizor
from config import Config, MeasureTracker
from requests import post

if __name__ == '__main__':
    # Process passed arguments
    parser = argparse.ArgumentParser()
    parser.add_argument('--config', help='Config file to be used')
    args = parser.parse_args()

    config_path = None if not args.config else args.config

    # Initialize classes
    config = Config(file_path=config_path)
    tracker = MeasureTracker(file_path=f'lm_{config.country}_{config.province}_{config.city}.conf.json')

    data = getPollutionData(config.iqair_api_key, config.country, config.province, config.city)
    info = Visualizor(data, True)

    if info.aqi != tracker.last_measure:
        info.generateImage()
        info.generateCaption()
        post(f'https://api.telegram.org/bot{config.telegram_api_key}/sendPhoto', data={'chat_id': config.telegram_chat_id, 'caption': info.caption, 'disable_notifications': True}, files={'photo': info.imageb})
        tracker.setMeasure(info.aqi)

misc.py
""" Aloodak
    Copyright (C) 2021  Frowzy et al

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.

    For copyright related issues, contact frowzyispenguin<at>riseup.net
"""
import pytz
from datetime import datetime, time
import jdatetime

jdatetime.set_locale("fa_IR")

def now():
    tz = pytz.timezone('Asia/Tehran') 
    tehran_now = datetime.now(tz)    
    return tehran_now.strftime("%H:%M")


def today():
    today = jdatetime.datetime.now(pytz.timezone("Asia/Tehran")).strftime("%A %y/%m/%d").split()
    date = today[-1]
    today.pop(-1)
    day = " ".join(i.strip() for i in today)
    return [day, date]


def toPersianNumerics(string: str):
    digits = {'1':'۱', '2':'۲', '3':'۳', '4':'۴', '5':'۵', '6':'۶', '7':'۷', '8':'۸', '9':'۹', '0':'۰'}
    string = str(string) # In case a number is being passed

    # Map the string into a list, replace a character if there's a digit
    chars = list(map(lambda x: digits[x] if x.isdecimal() else x,list(string)))
    
    # Join the mapped string back together and return a string
    return ''.join([str(x) for x in chars]) 

config.py
""" Aloodak
    Copyright (C) 2021  Frowzy et al

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.

    For copyright related issues, contact frowzyispenguin<at>riseup.net
"""
import json
import os
class Config:
    def __init__(self, file_path = None, ):
        if not file_path:
            self.file_path = 'config.conf.json'
        else:
            self.file_path = file_path
        self._read()

    def _read(self):
        with open(self.file_path, 'r') as file:
            self._config = json.load(file)
        
        for key, value in self._config.items():
            setattr(self, key, value)

class MeasureTracker(Config):
    
    def __init__(self, file_path="last_measure.conf.json"):
        self.file_path = file_path

        if not os.path.exists(self.file_path):
            self.setMeasure(-1)
        
        super().__init__(self.file_path)

    def setMeasure(self, measure: int):
        with open(self.file_path, 'w+') as file:
            json.dump({'last_measure': measure}, file)


aloodak.py
import jdatetime
import requests
import rtl
from bs4 import BeautifulSoup as bs
from PIL import Image, ImageDraw, ImageFont
from datetime import date, datetime
import os
import pytz
jdatetime.set_locale("fa_IR")

# for replacing english digits with persian ones
digits = {'1':'۱', '2':'۲', '3':'۳', '4':'۴', '5':'۵', '6':'۶', '7':'۷', '8':'۸', '9':'۹', '0':'۰'}
statusbar = {
    "Hazardous": "خطر اضطراری",
    "Very Unhealthy": "خیلی ناسالم",
    "Unhealthy": "ناسالم",
    "Unhealthy for Sensitive Groups": "ناسالم برای گروه های حساس",
    "Moderate": "سالم",
    "Good": "پاک"
}

# defining textcolor / light/dark colors has calculated by hps
color = {
    "Hazardous": (104,62,81),
    "Very Unhealthy": (99,70,117),
    "Unhealthy": (175,44,59),
    "Unhealthy for Sensitive Groups": (178,88,38),
    "Moderate": (165,127,35),
    "Good": (113,139,58)
}


def now():
    tz = pytz.timezone('Asia/Tehran') 
    tehran_now = datetime.now(tz)    
    return tehran_now.strftime("%H:%M")


def today():
    today = jdatetime.datetime.now(pytz.timezone("Asia/Tehran")).strftime("%A %y/%m/%d").split()
    date = today[-1]
    today.pop(-1)
    day = " ".join(i.strip() for i in today)
    return [day, en2per(date)]


def en2per(string):
    # it parses string in list for acessing to string charachters
    chars = list(map(lambda x: digits[x] if x.isdecimal() else x,list(string)))
    # it returns an string of above list elements wich joined
    return ''.join([str(x) for x in chars]) 


class aloodak:
    def __init__(self):
        pass
    def parser(country,province,city):
        # request webserver and parsing data
        data = {}
        key = "PUT YOU IQAIR API KEY HERE" 
        url = f"http://api.airvisual.com/v2/city?city={city}&state={province}&country={country}&key={key}"
        jsondata = requests.get(url).json() #json aqi and weather record
        soup = bs(requests.get(f"https://www.iqair.com/{country}/{city}", "hmtl.parser") #webpage aqi and weather record

        data['status'] = soup.find_all('span', attrs={"class":"aqi-status__text"})[0].text #air status
        data['aqi'] = en2per(jsondata['data']['current']['pollution']['aqius']) #aqi
        data['pm'] = en2per(soup.find_all("tbody", attrs={"_ngcontent-airvisual-web-c39":""})[4].find_all('td')[2].text) #pm2.5
        data['temperature'] = en2per(jsondata['data']['current']['weather']['tp']) #temperature
        data['humidity'] = en2per(j['data']['current']['weather']['hu']) #humidity
        return data


class info_maker():
    def __init__(self,data):
        self.imgdir = f"badges/{data['status']}.png"
        self.font = ImageFont.truetype("Sahel-Black.ttf", 140) # rate font style
        self.temperature = data['temperature']
        self.humidity = data['humidity']
        self.pm = data['pm']
        self.aqi = str(data['aqi']) # pollution rate
        self.status = statusbar[data['status']] # pollution status
        self.color = color[data['status']]

    def draw(self): 
        self.image = Image.open(self.imgdir) # opening image
        self.draw_object = ImageDraw.Draw(self.image) # loafing draw module 
        # wrting data over image
        if len(self.aqi) == 3:
            self.draw_object.text((100,100),self.aqi,self.color,font=self.font)
        elif len(self.aqi) == 2:
            self.draw_object.text((130,100),self.aqi,self.color,font=self.font)
        elif len(self.aqi) == 1:
            self.draw_object.text((150,100),self.aqi,self.color,font=self.font) 
        # saving image
        self.image.save("report.png")
        return None

    def checksum(self):
        return os.popen("sha1sum report.png").read().split()[0]

    def cpation(self):
        items = {"🔹شاخص آلودگی هوا : " : self.aqi,
                 "🔹وضعیت هوا : " : self.status,
                 "💧 رطوبت هوا : ":self.humidity,
                 "🌡 دمای هوا : ":self.temperature,
                 "💨 غلظت ذرات معلق در هوا : ":self.pm,
                 "🗓تاریخ : " : ' - '.join([x for x in today()]),
                 "🕓ساعت : " : en2per(now())
                }
        with open("report.txt","w") as foo :
            caption = [(item+items[item]) for item in items]
            caption = '\n'.join(x for x in caption)
            foo.write(caption)

